src/mlpack/methods/ann/vanilla_rbm_impl.hpp

```diff
@@ -0,0 +1,232 @@
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_VANILLA_RBM_IMPL_HPP
+#define MLPACK_METHODS_ANN_VANILLA_RBM_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "vanilla_rbm.hpp"
+
+#include <mlpack/core.hpp>
+#include <mlpack/prereqs.hpp>
+#include <mlpack/core/math/random.hpp>
+
+
+#include "layer/layer.hpp"
+#include "layer/base_layer.hpp"
+
+#include "init_rules/gaussian_init.hpp"
+#include "init_rules/random_init.hpp"
+#include "activation_functions/softplus_function.hpp"
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+// Intialise a linear layer and sigmoid layer
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::RBM(
+    arma::mat predictors,
+    InitializationRuleType initializeRule,
+    VisibleLayerType visible,
+    HiddenLayerType hidden,
+    const size_t numSteps,
+    const bool persistence):
+    visible(visible),
+    hidden(hidden),
+    initializeRule(initializeRule),
+    numSteps(numSteps),
+    persistence(persistence),
+    reset(false)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::Reset()
+{
+  // Since we are dealing with a static stucture no need for visitor
+  // Reset the parameters of all the layers
+  size_t weight = 0;
+  weight+= visible.Parameters().n_elem;
+  parameter.set_size(weight, 1);
+  parameter.zeros();
+  initializeRule.Initialize(parameter, parameter.n_elem, 1);
+  visible.Parameters() = arma::mat(parameter.memptr(), weight, 1, false, false);
+  hidden.Parameters() = arma::mat(parameter.memptr(), weight, 1, false, false);
+
+  visible.Reset();
+  hidden.Reset();
+  reset = true;
+}
+
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+template<template<typename> class Optimizer>
+void RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    Train(const arma::mat& predictors, Optimizer<NetworkType>& optimizer)
+{
+  numFunctions = predictors.n_cols;
+  this->predictors = std::move(predictors);
+
+  if (!reset)
+  {
+    Reset();
+  }
+
+  // Train the model.
+  Timer::Start("rbm_optimization");
+  optimizer.Optimize(parameter);
+  Timer::Stop("rbm_optimization");
+}
+
+ /* 
+  * This function calculates
+  * the free energy of the model
+  */
+template<typename InitializationRuleType, typename VisibleLayerType,
+    typename HiddenLayerType>
+double RBM<InitializationRuleType, VisibleLayerType, HiddenLayerType>::
+    FreeEnergy(arma::mat input)
```

Comment:
Why did you remove the reference? You needn't to copy the matrix since the function doesn't change that.